pyamg/strength.py

```diff
@@ -175,37 +182,81 @@ def classical_strength_of_connection(A, theta=0.0, norm='abs'):
     else:
         blocksize = 1
 
-    if not sparse.isspmatrix_csr(A):
-        warn('Implicit conversion of A to csr', sparse.SparseEfficiencyWarning)
-        A = sparse.csr_matrix(A)
-
     if (theta < 0 or theta > 1):
         raise ValueError('expected theta in [0,1]')
 
-    Sp = np.empty_like(A.indptr)
-    Sj = np.empty_like(A.indices)
-    Sx = np.empty_like(A.data)
+    # Block structure considered before computing SOC
+    if (block == 'block') or sparse.isspmatrix_bsr(A):
+        R, C = A.blocksize
+        if (R != C) or (R < 1):
+            raise ValueError('Matrix must have square blocks')
+
+        N = int(A.shape[0] / R)
+
+        # SOC based on maximum absolute value element in each block
+        if norm == 'abs':
```

Comment:
Just wanna make sure I understand.  Will data be positive for the 'abs' case and 'fro' case, but possibly negative for the 'min' case?